File: Study_Plans/leetcode-75/go/1372.go
package main


type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func longestZigZag(root *TreeNode) int {
    maxZigZag := 0
    var rec func (root *TreeNode, dirIsLeft bool, pathLen int)

    rec = func(root *TreeNode, dirIsLeft bool, pathLen int) {
        if root == nil {
            return 
        }

        maxZigZag = max(maxZigZag, pathLen)

        if dirIsLeft {
            rec(root.Left, false, pathLen + 1)
            rec(root.Right, true, 1)
        } else {
            rec(root.Right, true, pathLen + 1)
            rec(root.Left, false, 1)
        }

        return
    }

    rec(root, true, 0)
    rec(root, false, 0)

    return maxZigZag
}

File: Study_Plans/leetcode-75/1372.cc
#include <bits/stdc++.h>
#include <climits>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

class Solution {
    public:
        int longestZigZag(TreeNode *root) {
            int maxZigZag = 0;

            auto rec = [&](TreeNode *root, bool dirIsLeft, int pathLen, auto &&rec) {
                if (!root)
                    return;

                maxZigZag = max(maxZigZag, pathLen);

                if (dirIsLeft) {
                    rec(root->left, false, pathLen + 1, rec);
                    rec(root->right, true, 1, rec);
                } else {
                    rec(root->right, true, pathLen + 1, rec);
                    rec(root->left, false, 1, rec);
                }

                return;
            };

            rec(root, false, 0, rec);
            rec(root, true, 0, rec);

            return maxZigZag;

        }
};
